docs/spec/ibc/queues.md

```diff
@@ -0,0 +1,160 @@
+
+**head** &#8658; **i**: _q<sub>head</sub>_
+
+**tail** &#8658; **i**: _q<sub>tail</sub>_
+
+Based upon this needed functionality, we define a set of keys to be stored in the merkle tree, which allows us to efficiently implement and prove any of the above queries.
+
+**Key:** _(queue name, [head|tail|index])_
+
+The index is stored as a fixed-length unsigned integer in big endian format, so that the lexicographical order of the byte representation of the key is consistent with their sequence number. This allows us to quickly iterate over the queue, as well as prove the content of a packet (or lack of packet) at a given sequence. _head_ and _tail_ are two special constants that store an integer index, and are chosen such that their serialization cannot collide with any possible index.
+
+A message queue is simply a set of serialized packets stored at predefined keys in a merkle store, which can produce proofs for any key. Once a packet is written it must be immutable (except for deleting when popped from the queue). That is, if a value _v_ is written to a queue, then every valid proof _M<sub>k,v,h </sub>_ must refer to the same _v_. This property is essential to safely process asynchronous messages.
+
+Every IBC implementation must provide a protected subspace of the merkle store for use by each queue that cannot be affected by other modules.
+
+### 3.2   Naming Queues
+
+As mentioned above, in order for the receiver to unambiguously interpret the merkle proofs, we need a unique, deterministic, and predictable key in the merkle store for each message in the queue. We explained how the indexes are generated to provide each message in a queue a unique key, and mentioned the need for a unique name for each queue.
+
+The queue name must be unambiguously associated with a given connection to another chain, so an observer can prove if a message was intended for chain A or chain B. In order to do so, upon registration of a connection with a remote chain, we create two queues with different names (prefixes).
+
+* _ibc:<chain id of A>:send_ - all outgoing packets destined to chain A
+* _ibc:<chain id of A>:receipt_ - the results of executing the packets received from chain A
+
+These two queues have different purposes and store messages of different types. By parsing the key of a merkle proof, a recipient can uniquely identify which queue, if any, this message belongs to. We now define _k =_ _(remote id, [send|receipt], index)_. This tuple is used to route and verify every message, before the contents of the packet are processed by the appropriate application logic.
+
+### 3.3   Message Contents
+
+Up to this point, we have focused on the semantics of the message key, and how we can produce a unique identifier for every possible message in every possible connection. The actual data written at the location has been left as an opaque blob, put by providing some structure to the messages, we can enable more functionality.
+
```

Comment:
Is payload blob or we assume some structure?

Reply:
Payload is a blob, clarified this in a few places.